src/Compiler/Utilities/illib.fs

```diff
@@ -78,6 +81,8 @@ module internal PervasiveAutoOpens =
         member inline x.EndsWithOrdinalIgnoreCase value =
             x.EndsWith(value, StringComparison.OrdinalIgnoreCase)
 
+    let inline isDigit (c: char) = (uint)(c - '0') <= (uint)('9' - '0');
```

Comment:
Can you please make a comment why do we chose to use custom function instead of the one from Char?

Reply:
There is another one https://github.com/dotnet/fsharp/pull/16066/files#diff-afc3ed6b692f9f687b04f4f72e14b3d2592245ca7a334476c07c162838d90304R87 , but in FSharp.Core and neither is accessible there https://github.com/dotnet/fsharp/blob/fb39b9440dd27f5ed093e0617cdce49d9dbffa3c/src/Compiler/Utilities/sformat.fs#L1531C48-L1531C60
Any suggestions?

Reply:
I think it's fine to leave those. I'm mostly asking why do we use custom one in compiler now instead of BCL one now?

Solely for the purpose of inlining?

Reply:
No, the whole point of this PR is that `Char.IsDigit` (see [impl](https://github.com/dotnet/runtime/blob/c92caa419873bbb5c51997c0f613ea5b5b1d22ec/src/libraries/System.Private.CoreLib/src/System/Char.cs#L317)) is testing for Unicode definition of 'Digit'. So it returns `true` not only for `0-9`, but also for the whole bunch of [other code points](https://www.fileformat.info/info/unicode/category/Nd/list.htm). But the code doesn't account for that i.e. https://github.com/dotnet/fsharp/pull/16066/files#diff-afc3ed6b692f9f687b04f4f72e14b3d2592245ca7a334476c07c162838d90304R160 `'۹' - '0' = 1737`
From performance perspective it's clearly win -- this should compile to only 3 instructions (add,cmp,setb) and eliminates branch, but I assume this will be completely unnoticeable.

Reply:
The other part is switch to char overloads where possible (they are not culture sensitive) and explicitly specify `Ordinal` with string overloads (default is `CurrentCulture`).
\+ bunch of `ToLower` -> `ToLowerInvariant` for deterministic compilation regardless user culture.

Reply:
I see, you mentioned that it's technically a breaking change, which is technically true. Can you think of any specific examples it may break?

Reply:
Yes, it's presented in OP.
Currently fsc accepting non ascii digits as valid printf width/precession specifiers (both as literal and as `*` arg) and makes non-nonsensical calculations with them. Now it will error out. Most other usages of `Char.IsDigit` are mostly in IL parsing.

[This check](https://github.com/dotnet/fsharp/pull/16066/files#diff-a4284a03583b6abfea98b69073e5409f3fe132a6caf7b2700e6e26ee6430e55cL3248) for example fails for scripts without cases (Hebrew, most Asian langs).
Non invariant ToUpper/Lower with tr_TR and az_AZ locales trashes letter `i` (it doesn't roundtrip).
Culture sensitive path compassion [here](https://github.com/dotnet/fsharp/pull/16066/files#diff-7ab39733da4162be47a5ab74224f8a7b8e7f129e2b67538534f0a0135f02193eL1320) may resolve to different files.

Reply:
> Yes, it's presented in OP.
> 
> Currently fsc accepting non ascii digits as valid printf width/precession specifiers (both as literal and as `*` arg) and makes non-nonsensical calculations with them. Now it will error out. Other usages of `Char.IsDigit` are mostly in IL parsing.
> 
> 
> 
> [This check](https://github.com/dotnet/fsharp/pull/16066/files#diff-a4284a03583b6abfea98b69073e5409f3fe132a6caf7b2700e6e26ee6430e55cL3248) for example fails for scripts without cases (Hebrew, most Asian langs).
> 
> Non invariant ToUpper/Lower with tr_TR and az_AZ locales trashes letter `i` (it doesn't roundtrip).
> 
> Culture sensitive path compassion [here](https://github.com/dotnet/fsharp/pull/16066/files#diff-7ab39733da4162be47a5ab74224f8a7b8e7f129e2b67538534f0a0135f02193eL1320) may resolve to different files.

Oh, so most of them are not the most legit use cases. This type of breaking change is ok, I think and we can treat it as a fix.